src/lib/reasoners/bitv.ml

```diff
@@ -692,21 +692,39 @@ module Shostak(X : ALIEN) = struct
         |_ -> assert false
       in List.fold_left f_aux (List.hd s_l)(List.tl s_l)
 
-    let slice_var var s1 =
-      let s2 = var.sz - s1 in
-      match var.bv with
-      |S_Cte _ -> {var with sz = s1},{var with sz = s2},None
-      |S_Var v ->
-        let (fs,sn,tr) = match v.sorte with
-          |A -> (fresh_var A), (fresh_var A), A
-          |B -> (fresh_var B), (fresh_var B), B
-          |C -> (fresh_var C), (fresh_var C), C
-        in {bv = S_Var fs; sz = s1},{bv = S_Var sn; sz = s2},Some tr
+    (* [slice_var var pat_hd pat_tl] slices a variable to make it fit the
+       pattern [pat_hd :: pat_tl].
+       Returns the list of created fresh variables, the left over pattern and
+       the sort of the variable (if it is not a constant). *)
+    let slice_var var pat_hd pat_tl =
+      let mk, vsz, tr =
+        match var.bv with
+        | S_Cte _ -> (fun sz -> { var with sz }), var.sz, None
+        | S_Var { sorte = A; _ } ->
+          (fun sz -> { bv = S_Var (fresh_var A); sz }), var.sz, Some A
+        | S_Var { sorte = B; _ } ->
+          (fun sz -> { bv = S_Var (fresh_var B); sz }), var.sz, Some B
+        | S_Var { sorte = C; _ } ->
+          (fun sz -> { bv = S_Var (fresh_var C); sz }), var.sz, Some C
```

Comment:
`vsz` is not needed in the match, it is always equal to `var.sz`

Reply:
Also I think this can be simplified:

```ocaml
let mk, tr =
  match var.bv with
  | S_Cte _ -> (fun sz -> { var with sz }), None
  | S_Var { sorte; _ } -> (fun sz -> { bv = S_Var (fresh_var sorte); sz }), Some sorte
in
...
```

Reply:
Done